Make sure SSO logins can handle duplicate usernames

etnoy opened the pull request:
In the SSO login context, it is possible that two users get the same initial usernames if their full legal names are the same. This PR handles this case by adding a number at the end of the username before allowing the user to change their username.

etnoy committed:
Cleanup for PR

etnoy committed:
Restore modified database test file

etnoy committed:
Merge branch 'dev#510' of github.com:etnoy/SecurityShepherd into dev

etnoy committed:
Add setup IT

etnoy committed:
Initialize string buffer

etnoy committed:
Remake setup IT to test core database

etnoy committed:
Use commons IO for auth file checking

etnoy committed:
Set all db properties before setup

etnoy committed:
Add teardown to dbprops

etnoy committed:
Double negation fixed

etnoy committed:
Make sure params are added to the right place

etnoy committed:
Check return code of failing setup page

etnoy committed:
Improve teardown

etnoy committed:
Merge branch 'dev' of github.com:OWASP/SecurityShepherd into dev

etnoy committed:
Merge branch 'dev#510' into dev

etnoy committed:
Rename db props to mysql props

etnoy committed:
Merge branch 'dev' of github.com:etnoy/SecurityShepherd into dev

etnoy committed:
Tests dynamically create db props file now

etnoy committed:
Update unit tests to create db file

etnoy committed:
All IT tests have to create mysql properties file

etnoy committed:
Hunting for null pointer in SetupIT

etnoy committed:
Disabling setup it tests for now

etnoy committed:
Add error handling for db connection issues

etnoy committed:
Add error handling for db connection issues

etnoy committed:
Merge branch 'dev#510' into dev

etnoy committed:
Add saml dependency and metadata jsp

etnoy committed:
Add demo saml files

etnoy committed:
Add test for upper and lowercase username

etnoy committed:
Add test for upper and lowercase username

etnoy committed:
Test for long passwords

etnoy committed:
Test for long passwords

etnoy committed:
Integration test for 160-char password

etnoy committed:
Integration test for 160-char password

etnoy committed:
Make sure IT tests dont' step on each others toes

etnoy committed:
Make sure IT tests dont' step on each others toes

etnoy committed:
Fix naming of testUtils

etnoy committed:
Add utf-8 username test

etnoy committed:
Test for non-latin passwords

etnoy committed:
Add argon2 dependency

etnoy committed:
Test for duplicate usernames

etnoy committed:
Add password change tests

etnoy committed:
Catch more generic SQL exception

etnoy committed:
Salted passwords with argon2

etnoy committed:
Fix syntax error in core sql

etnoy committed:
Reset userfind to upstream version

etnoy committed:
Compute hash correctly in userAuth

etnoy committed:
Add salted password to default sql

etnoy committed:
Merge branch 'dev#510' into dev#513

etnoy committed:
Fix invalid update password logic

etnoy committed:
Print debug info in authuser

etnoy committed:
Add debugging to userCreate

etnoy committed:
Install argon2 in travis

etnoy committed:
More specific exception handling in userAuth

etnoy committed:
Fix whitespace in travis.yml

etnoy committed:
Add argon2 install to unit tests

etnoy committed:
We need bionic because of libargon

etnoy committed:
Move dist definition

etnoy committed:
Don't throw exception if hash wasn't found

etnoy committed:
Handle user suspensions

etnoy committed:
Add very long username test

etnoy committed:
Don't print hashes and passwords in log

etnoy committed:
Ensure username is lowercase everywhere

etnoy committed:
More specific username length testing

etnoy committed:
Test username length with non-latin chars

etnoy committed:
Make username more case-insensitive

etnoy committed:
Use utf8 in database

etnoy committed:
Fix problematic non latin tests

etnoy committed:
More specific test for non-latin username

etnoy committed:
Slightly shorter non-latin username

etnoy committed:
Add db properties option

etnoy committed:
Add support for utf8mb4 for emoji goodness

etnoy committed:
Add emoji to unit tests

etnoy committed:
Big utf8 means varchar can only be 191

etnoy committed:
Separate testing for emoji usernames

etnoy committed:
Fix missing slash in test db props

etnoy committed:
Add more unicode stuff to mysql

etnoy committed:
Disable emoji test for now

etnoy committed:
Fix typo in sql injection database handler

etnoy committed:
Remove extra whitespace

etnoy committed:
Clear up some unused imports

etnoy committed:
Merge branch 'dev#513' into dev

etnoy committed:
Merge remote-tracking branch 'origin/dev' into dev

etnoy committed:
Ignore incomplete tests

etnoy committed:
Add sonarcloud properties

etnoy committed:
First attempt att registering users from SSO

etnoy committed:
Cleaning up exception handling in unit and IT tests

etnoy committed:
Disabling passwords for SSO logins

etnoy committed:
Refactoring tests to reduce bad exception handling

etnoy committed:
Handle SQL exceptions in test code

etnoy committed:
Removing more bad exception handling

etnoy committed:
Removed Exception e in testCSRF

etnoy committed:
CSRF unit test can now pass

etnoy committed:
Cleaned up most module IT tests

etnoy committed:
Cleanup test setups and fix missed user creation in OpenAllModules

etnoy committed:
More cleanup of IT exception handling

etnoy committed:
Only deploy if it's the official repo

etnoy committed:
Add coveralls

etnoy committed:
Add codecov

etnoy committed:
Refactoring travis yml

etnoy committed:
Call codecov from travis

etnoy committed:
Remove cobertura

etnoy committed:
Remove coveralls

etnoy committed:
Change jdk version

etnoy committed:
Remove oraclejdk8 as travis doesn't support it

etnoy committed:
Remove failing maving packages

etnoy committed:
Refactoring travis yml

etnoy committed:
Trying to reorder travis

etnoy committed:
Remove jdk versions for now

etnoy committed:
Remove unused parts of pom

etnoy committed:
Revert to former yml

etnoy committed:
Trying to revert pom.xml

etnoy committed:
Add dependencies

etnoy committed:
New travis again

etnoy committed:
Trying multiple jdks

etnoy committed:
Travis cleanup

etnoy committed:
Add user from SSO

etnoy committed:
Add SSO suspend tests

etnoy committed:
Fix invalid SSO user handling

etnoy committed:
Redone core.user table to fit SSO/SAML

etnoy committed:
Change db format for SSO

etnoy committed:
Set saml in quotes

etnoy committed:
Set field to null in ssoName

etnoy committed:
SSO default class handling

etnoy committed:
Fix typo in sso user creation

etnoy committed:
Check return value of userCreated

etnoy committed:
Update test for SSO suspension

etnoy committed:
Test user SSO delete

etnoy committed:
Start to handle temporary usernames

etnoy committed:
SAML can now change usernames

etnoy committed:
Can change username on the fly

etnoy committed:
Fix broken updateUsername test

etnoy committed:
Set default value for change username

etnoy committed:
VerifyTestUser ignored passwords, now fixed

etnoy committed:
General SSO cleanup, moved from jsp to servlets

etnoy committed:
Add sample saml properties

etnoy committed:
Try to handle more generic saml keys, configurable via external file

etnoy committed:
Cleaned up sso configuration, can now handle renames and robust sso on/off handling

etnoy committed:
Roles are now assigned via SAML

etnoy committed:
Update sso links

etnoy committed:
Fix db syntax

etnoy committed:
Db options are now mandatory

etnoy committed:
Test suites create db props with utf8 support

etnoy committed:
Dynamic creation of mongodb props file for tests

etnoy committed:
Disable emoji username test for now

etnoy committed:
Create parent directories for tests

etnoy committed:
Fix comments from @markdenihan

etnoy committed:
Revert to upstream travis yml

etnoy committed:
Merge branch 'dev#513' into saml-login

etnoy committed:
Setter and getter test for cheat sheet setting

etnoy committed:
Save cheat sheet status to db

etnoy committed:
Can now read cheat sheet status from db

etnoy committed:
Refactor cheat sheet methods

etnoy committed:
Install argon2 in travis

etnoy committed:
Merge branch 'saml-login' into dev#329

etnoy committed:
Correct syntax in cheat sheet load and save

etnoy committed:
Set and get module layout

etnoy committed:
Change tourny to tournament

etnoy committed:
Re-do nosql IT module setup

etnoy committed:
Merge branch 'saml-login' into dev#329

etnoy committed:
Set module layout to db

etnoy committed:
More specific tests for module layout db

etnoy committed:
More specific tests for module layout db 2

etnoy committed:
Feeback status saved to db

etnoy committed:
Save registration status to db

etnoy committed:
Save scoreboard status in db

etnoy committed:
Load and save scoreboard data from db

etnoy committed:
Module layouts working

etnoy committed:
Public scoreboard done

etnoy committed:
Fix failed scoreboard test

etnoy committed:
Fix failing feedback tests

etnoy committed:
Dont say failed unless failed

etnoy committed:
Add more debugging to feedback tests

etnoy committed:
Update enablefeedback setting in sql

etnoy committed:
More debugging for scoreboard IT failures

etnoy committed:
Log more data from load scoreboard status

etnoy committed:
Dont clear scoreboard class with open layout

etnoy committed:
Dont clear scoreboard class with admin layout

etnoy committed:
More debugging for null pointer tracing

etnoy committed:
Debug logging for db props creation in tests

etnoy committed:
Add countdown handling class

etnoy committed:
Debug mongo db loading

etnoy committed:
Show when mongo db is loaded

etnoy committed:
Even more debug logging

etnoy committed:
Hunting NPE in mongo

etnoy committed:
Even more debugs plz

etnoy committed:
Fixed the bug

etnoy committed:
Merge branch 'dev#329' into countdown

etnoy committed:
Wrote timestamp set get tests

etnoy committed:
Tests for countdownhandler

etnoy committed:
Improve debug logging for failing tests

etnoy committed:
Fix typo that removed timestamp

etnoy committed:
Countdown built into the stored procedures

etnoy committed:
More tests for countdown

etnoy committed:
Add start time to countdown

etnoy committed:
Validate start and end conditions

etnoy committed:
Add js countdown timer

etnoy committed:
Merge remote-tracking branch 'origin/dev' into dev#329

etnoy committed:
Fix merge

etnoy committed:
Merge branch 'dev#329' into countdown

etnoy committed:
Countdown shows on index page

etnoy committed:
Cleanup module IT tests

etnoy committed:
Start and Lock countdowns work

etnoy committed:
Remove unused imports

etnoy committed:
Handle NPE in GetModule

etnoy committed:
Code cleanup

etnoy committed:
Check for null moduleid

etnoy committed:
Remove most of fileinputproperties

etnoy committed:
Fix unit tests

etnoy committed:
Cleanup sql injection db loader

etnoy committed:
Can now override database settings from web UI

etnoy committed:
Some login cleanup

etnoy committed:
Set cookie path in saml

etnoy committed:
Remove debug info

etnoy committed:
Remove extra property from mongo

etnoy committed:
Make sure html is lowercase

etnoy committed:
Optimize performance in scoreboard

etnoy committed:
Optimize performance in registration logic

etnoy committed:
Show two countdowns

etnoy committed:
Merge github.com:OWASP/SecurityShepherd into dev#521

etnoy committed:
Replace rage faces with fruit

etnoy committed:
Spelling of crypto

etnoy committed:
Upgrade sql stored procedure to support db options

etnoy committed:
Only filter XSS3 twice

etnoy committed:
Change the admin session cookie to be easier to find

etnoy committed:
Correct the sso metadata logger argument

etnoy committed:
De-duplicate usernames!

etnoy committed:
Unit test for de-duplication of sso usernames

etnoy committed:
Add more tests to duplicate sso usernames

etnoy committed:
Fix bug in de-duplication

markdenihan commented:
@etnoy In the effort of condensing reviewing work, can #530 #528 #524 #523 #520 #519 and #518 be closed as it appears that this Pull Request is a combination of all in one?

etnoy commented:
> @etnoy In the effort of condensing reviewing work, can #530 #528 #524 #523 #520 #519 and #518 be closed as it appears that this Pull Request is a combination of all in one?

Sure, let's do it.

etnoy committed:
Fix utf-8 scoreboard usernames

etnoy committed:
Forgot to add assignments properties

etnoy committed:
Fix classID for SSO logins

etnoy committed:
Make more servlets utf-8 compatible

etnoy committed:
Cache module layout in memory

etnoy committed:
Cache cheat status in memory

etnoy committed:
Fix challenge connection urls

etnoy committed:
Set correct key in db options

etnoy committed:
Fix direct object bank registration

etnoy committed:
Set correct SHA2 lengths

etnoy committed:
Save default class info to db

etnoy committed:
Change log statement in setter

etnoy committed:
Load default class when doing account registration

etnoy committed:
Fix wrong order of submitted and expected keys in logs

etnoy committed:
Scoreboard can be viewed publicly

etnoy committed:
Allow a public scoreboard

Theadri1900 committed:
Fixing database bug in SQL injection 6

Fixed a bug where the database provided an incorrect answer key. 
_Because of bad luck with commit c57fed8 where all "256" values were replaced by "191"_

etnoy committed:
Merge pull request #1 from Theadri1900/Theadri1900-patch-2

Fixing database bug in SQL injection 6

etnoy commented:
Any input on this? I have more issues to work on but it gets messy when I can't start from the dev branch

markdenihan commented:
Hey Jonathan,

I'm hoping to get some spare time to get through this pull over the next
few days. I've started reviewing it but it's massive so it's taking a
while. Sorry for the delay

On Fri 20 Dec 2019, 9:38 p.m. Jonathan Jogenfors, <notifications@github.com>
wrote:

> Any input on this? I have more issues to work on but it gets messy when I
> can't start from the dev branch
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OWASP/SecurityShepherd/pull/531?email_source=notifications&email_token=AA2SVHV4S3BTA4XUQEKQQELQZU3M5A5CNFSM4JQPEAT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHOHUJQ#issuecomment-568097318>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2SVHS3E5GMTLKHOT2CFXLQZU3M5ANCNFSM4JQPEATQ>
> .
>

markdenihan commented on src/main/java/servlets/ChangeUsername.java:
```diff
@@ -0,0 +1,100 @@
+					log.debug("New username passed validation! Username Change gets the go ahead");
+					Setter.updateUsername(ApplicationRoot, userName, newUsername);
+					ses.setAttribute("ChangeUsername", "false");
+					ses.setAttribute("userName", newUsername);
```
newUsername should be escaped for html before it is put in the userName session attribute, similarly to how it is set in the index.html;
Encode.forHtml();

etnoy replied:
Good catch. Don't want to introduce an XSS

markdenihan commented on src/main/java/servlets/ChangeUsername.java:
```diff
@@ -0,0 +1,100 @@
+				if (validUsername) {
+
+					log.debug("New username passed validation! Username Change gets the go ahead");
+					Setter.updateUsername(ApplicationRoot, userName, newUsername);
```
Doesn't appear to be any flow other than the happy path here. If a user tries to update their username to one that exists already the SQL Schema will prevent that from happening. Doesn't appear to be anything that tells the user "that didn't work" in any scenario where the DB complains

etnoy replied:
There will actually be an error message popped back to the user because Setter.updateUsername throws a RuntimeException if the username is taken. This can be handled more cleanly, I agree, but for now I think this solution works (and is not worse than the shepherd codebase in general when dealing with exceptions)

markdenihan commented on src/main/java/servlets/ChangeUsername.java:
```diff
@@ -0,0 +1,100 @@
+				log.debug("New username: " + newUsername);
+
+				boolean validUsername = false;
+				validUsername = newUsername.length() > 3 && newUsername.length() <= 32;
```
Should be either;
changed to use Validate.isValidUser(username, password) with a hardcoded password that passes the password validation
Changed to use Validate.isValidUser(username), a new method that just validates usernames

so that if we change the rules in the future we don't forget that this needs to be updated as well

etnoy replied:
Good point, although I think this should be handled in a separate feature request. ChangePassword.java does the same type of length test without breaking it out in a separate function. I suggest we leave as-is for now.

markdenihan commented on src/main/java/utils/XssFilter.java:
```diff
@@ -156,7 +156,7 @@ public static String levelThree (String input)
 		for(int h = 0; h < FindXSS.javascriptTriggers.length; h++)
 		{
-			for(int i = 0; i <= 3; i++)
+			for(int i = 0; i <= 1; i++)
```
Looks like this change was unintended. Can it be reverted?

etnoy replied:
See #527. I strongly advocate for keeping this change as I've seen many students (at this point we have had >300 people use shepherd) struggle with this specific filter.

markdenihan commented on src/main/java/servlets/module/challenge/SessionManagement4.java:
```diff
@@ -98,7 +98,7 @@ public void doPost (HttpServletRequest request, HttpServletResponse response)
 						log.debug("Guest Session Detected");
 					}
-					else if (decodedCookie.equals("0000000000000021")) //Admin Session
+					else if (decodedCookie.equals("0000000000000009")) //Admin Session
```
Any reason why this was changed from 0000000000000021 to 0000000000000009? Makes the level a bit easier?
If this change sticks it's rank/points should be reassessed

etnoy replied:
Please see #529. Agreed on changing rank/points if needed.

markdenihan commented on src/main/resources/i18n/cheatsheets/solutions.properties:
```diff
@@ -33,7 +33,7 @@ fccf8e4d5372ee5a73af5f862dc810545d19b176.solution = This challenge does not requ
 c7ac1e05faa2d4b1016cfcc726e0689419662784.solution = View the source of this challenge and inspect the JavaScript at the end of the page. You can deobfusticate the JavaScript with an on line web tool, give it a Google. The script contains two functions, and one of them is for a HTML element that is not in the level. This function is an admin function and reveals a URL very similar to the user func ...
 b3cfd5890649e6815a1c7107cc41d17c82826cfa.solution = There are a number of ways to defeat the crypto and get the encryption key in this challenge. The quickest way is to submit base64 encoded spaces. The crypto XOR's the spaces with the key and returns the resultant 'cipher text which is the encryption key.
-ced925f8357a17cfe3225c6236df0f681b2447c4.solution = Users must discover the session id for this sub application is very weak. The default session ID for a guest will be 00000001 base64'd twice. The admins session ID is TURBd01EQXdNREF3TURBd01EQXlNUT09 (00000021 when decoded).
+ced925f8357a17cfe3225c6236df0f681b2447c4.solution = Users must discover the session id for this sub application is very weak. The default session ID for a guest will be 00000001 base64'd twice. The admin's session ID is TURBd01EQXdNREF3TURBd01EQXdPUT09 (0000000000000009 when decoded).
```
Not sure why this is being changed?

etnoy replied:
Again, see #529

markdenihan reviewed (changes requested):
Sorry for the delay here, but I'm through all of the changes at last.

Please keep any changes in this merge request to minimize the effort of re-reviewing. I think it is only minor stuff that came up

etnoy commented:
Sorry for not getting back to this yet, it's on its way! Thanks for looking into my changes.

etnoy committed:
Encode new username before reading, prevent XSS

etnoy committed:
Merge branch 'dev' into duplicate-user

etnoy committed:
Bump argon2 version down to 2.2

etnoy commented:
I have updated the code as per the above comments. Should be ready to merge now.

etnoy commented:
This PR also fixes #557 and #552

markdenihan commented:
@etnoy Thanks so much for this work. It's really appeciated!

etnoy commented:
That's nice to hear. Glad to help :)
